src/language-js/print/decorators.js

```diff
@@ -0,0 +1,93 @@
+"use strict";
+
+const {
+  isNonEmptyArray,
+  hasNewlineInRange,
+  hasNewline,
+  getLast,
+} = require("../../common/util");
+const {
+  builders: { line, hardline, join, breakParent, group },
+} = require("../../document");
+const { locStart, locEnd } = require("../loc");
+const { getParentExportDeclaration } = require("../utils");
+
+function printClassMemberDecorators(path, options, print) {
+  const node = path.getValue();
+  return group([
+    join(line, path.map(print, "decorators")),
+    hasNewlineBetweenOrAfterDecorators(node, options) ? hardline : line,
+  ]);
+}
+
+function printDecoratorsBeforeExport(path, options, print) {
+  // Export declarations are responsible for printing any decorators
+  // that logically apply to node.declaration.
+  return [
+    join(hardline, path.map(print, "declaration", "decorators")),
+    hardline,
+  ];
+}
+
+function printDecorators(path, options, print) {
+  const node = path.getValue();
+  const { decorators } = node;
+
+  if (
+    !isNonEmptyArray(decorators) ||
+    // If the parent node is an export declaration and the decorator
+    // was written before the export, the export will be responsible
+    // for printing the decorators.
+    hasDecoratorsBeforeExport(path.getParentNode())
+  ) {
+    return;
+  }
+
+  const shouldBreak =
+    node.type === "ClassExpression" ||
+    node.type === "ClassDeclaration" ||
+    hasNewlineBetweenOrAfterDecorators(node, options);
+
+  return [
+    getParentExportDeclaration(path)
+      ? hardline
+      : shouldBreak
+      ? breakParent
+      : "",
+    join(line, path.map(print, "decorators")),
+    line,
+  ];
+}
+
+function hasNewlineBetweenOrAfterDecorators(node, options) {
+  const start = locEnd(node.decorators[0]);
+  const end = locEnd(getLast(node.decorators));
+  return (
+    hasNewlineInRange(options.originalText, start, end) ||
```

Comment:
@thorn0 I think this function is not right,

**Prettier 2.2.1**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAucAbAhgZ0wAgII7AA6UOOAAgCZyQBO6MEtAFMAL4CUOADrRN3FowAnjgC8OAIwBuEmxIkwGbPkIkyVGkwZNW6sjn2cefAUNESZckABoQ-GAEtomZKHS0+AdwAKHhK4o6Khe6MKudgBG9GAA1nAwAMrc6GCOUADmyDC0AK5wdnAAtpFwlNSUADLombnoGXAAYkxFDE6ZyCDouYy2IAAWMEWoAOr9jvCYKWBwiQETjgBuE8KdYNh96ZiCMD70Ga3IAGbB23YAVpgAHgBCMfFJ6EVwlelwx6cFIJdXiekZqDgAEVchB4B9UGcQClaNtaJ1IuhSqg+rx0jARo5KDB+sgABwABjsvAg2xG9G4nV4cDhi3edgAjqD4Ht+IEupgALRQOBlMp9WhwJmOQV7eqHJAnSFfbZFRzZPIy-6AkFg96Sz52GBIzHY3FIABMWvojlQ-wAwhAihKQDSAKx9XLbAAqSMCUqhi3yAEkoNRYIkwLRHNwYHg-YkRICIds2GwgA)
<!-- prettier-ignore -->
```sh
--parser babel
```

**Input:**
<!-- prettier-ignore -->
```jsx
class A {
  @decorator({}) property = 1;
}

class A {
  @decorator({
    
  }) property = 1;
}
```

**Output:**
<!-- prettier-ignore -->
```jsx
class A {
  @decorator({}) property = 1;
}

class A {
  @decorator({})
  property = 1;
}

```